### DISCUSSION TITLE
Anonymous Types & Deconstructor

### DISCUSSION THREAD

**Discussion by stidsborg** [dotnet/csharplang#6349] [Language Ideas]
When coding today I realized that anonymous types do not have an associated deconstructor. Thus, the following is not allowed:

```csharp
var (name, age) = new { Name = "Peter", Age = 35 };
```

I think adding this could often simplify code, as anonymous types are frequently used to pass a subset of information through a chain of LINQ operators. Thus, the anonymous type by itself is just a wrapper for distinct pieces of information.

For instance, the following situation arose today in my open source framework Cleipnir:
```csharp
var (status, postponedUntil) = await store.GetFunction(functionId).Map(sf => new {sf?.Status, sf?.PostponedUntil });
```

An alternative could be using a tuple. However, the IDE suggests names (item1, item2). Also it is suboptimal if a chain of LINQ operators must end with a Select converting the anonymous type into a tuple. 

I realize that this might result in an increased byte code size if implemented naively. I am hoping that this can be solved by some great compiler people.

Thanks in advance
/T

**Comment by jnm2** [comment]
It's worth discussing whether positional deconstruction makes sense for anonymous types since they are constructed nominally as in `{ Name = "Peter", Age = 35 }` rather than being constructed positionally as in `("Peter", 35)`.

One thing that bypasses that discussion is the championed work on allowing nominal deconstruction: https://github.com/dotnet/csharplang/issues/4082

```cs
var { Name: name, Age: age } = new { Name = "Peter", Age = 35 };

var { Status: status, PostponedUntil: postponedUntil } = await store.GetFunction(functionId).Map(sf => new {sf?.Status, sf?.PostponedUntil });
```

**Reply by HaloFour** [reply]
You beat me to it, I do think nominal deconstruction makes more sense here.  That's not to say that positional deconstruction couldn't work too, given that anonymous types sorta walk the line between positional and nominal, and to support it wouldn't require changing how the anonymous types are emitted given that those details are effectively private anyway and the compiler already knows what order the members are in based on the previous declaration, regardless of the actual implementation details of the type that is emitted.

```cs
var (name, age) = new { Name = "Peter", Age = 35 };

// compiles to
// (assuming the compiler wouldn't just optimize it away, like it does with such tuple assignments/deconstruction)
var $temp = new $AnonymousType1() { Name = "Peter", Age = 35 };
var name = $temp.Name;
var age = $temp.Age;
```

**Reply by CyrusNajmabadi** [reply]
> You beat me to it, I do think nominal deconstruction makes more sense here. 

Yup.  And i would expect this to just 'fall out'.  Similar to how `with` falls out for anonymous types.   It is good to make a note of this so we don't accidentally miss this as part of any deconstruction work we do though!

**Reply by stidsborg** [reply]
Thanks for the quick feedback
Interesting concept nominal deconstruction - did not know about it - will read up on it. 

Just a quick thought though. As a developer I quite like the minimum verboseness of
```csharp
var (name, age) = new { Name = "Peter", Age = 35 };
```
versus
```csharp
var (Name = name, Age = age) = new { Name = "Peter", Age = 35 };
```

It might be that an IDE would generate the deconstruction automatically - in which case it is better - but I am still not a fan of the duplication when reading the nominal deconstructed code. 

Also I noted that the compiler does not allow this:
```csharp
var p = new { Name = "Peter", Age = 35 };
var p2 = new { Age = 35, Name = "Peter" };
p = p2;
```

So it seems the compiler underneath does have a positional understanding of an anonymous type. However, I agree that this might be a compiler implementation detail. 

Anyways, many thanks for considering this and the feedback.

Best,
/T
